filters/include/pcl/filters/experimental/voxel_grid.h

```diff
@@ -0,0 +1,524 @@
+      const std::uint8_t* pt_data = reinterpret_cast<const std::uint8_t*>(&pt);
+      float distance_value = 0;
+      memcpy(&distance_value,
+             pt_data + filter_fields_[filter_field_idx_].offset,
+             sizeof(float));
+
+      if (getDerived()->getFilterLimitsNegative()) {
+        // Use a threshold for cutting out points which inside the interval
+        if ((distance_value < filter_limit_max_) &&
+            (distance_value > filter_limit_min_))
+          return;
+      }
+      else {
+        // Use a threshold for cutting out points which are too close/far away
+        if ((distance_value > filter_limit_max_) ||
+            (distance_value < filter_limit_min_))
+          return;
+      }
+    }
+
+    const std::size_t h = hashPoint(pt);
+
+    const bool downsample_all_data = getDerived()->getDownsampleAllData();
+    if (!downsample_all_data)
+      grid_[h].coord_centroid += pt.getVector4fMap();
+    else
+      grid_[h].centroid.add(pt);
+
+    grid_[h].num_pt++;
+  }
+
+  inline boost::optional<PointT>
+  filterGrid(GridIterator grid_it)
+  {
+    auto& voxel = grid_it->second;
+    const std::size_t min_points_per_voxel =
+        getDerived()->getMinimumPointsNumberPerVoxel();
+    if (voxel.num_pt >= min_points_per_voxel) {
+
+      const bool save_leaf_layout = getDerived()->getSaveLeafLayout();
+      if (save_leaf_layout)
+        leaf_layout_[grid_it->first] = num_voxels_++;
+
+      PointT centroid;
+      const bool downsample_all_data = getDerived()->getDownsampleAllData();
+      if (!downsample_all_data) {
+        voxel.coord_centroid.array() /= voxel.num_pt;
+        centroid.getVector4fMap() = voxel.coord_centroid;
+      }
+      else {
+        voxel.centroid.get(centroid);
+      }
+      return centroid;
+    }
+
+    return boost::none;
+  }
+
+protected:
+  /** \brief The filter name. */
+  std::string filter_name_;
+
+  /** \brief The size of a leaf. */
+  Eigen::Vector4f leaf_size_;
+
+  /** \brief Internal leaf sizes stored as 1/leaf_size_ for efficiency reasons. */
+  Eigen::Array4f inverse_leaf_size_;
+
+  /** \brief The leaf layout information for fast access to cells relative to current
+   * position **/
+  std::vector<int> leaf_layout_;
+
+  /** \brief The minimum and maximum bin coordinates, the number of divisions, and the
+   * division multiplier. */
+  Eigen::Vector4i min_b_, max_b_, div_b_, divb_mul_;
+
+  /** \brief The iterable grid object for storing information of each fraction of space
+   * in the filtering space defined by the grid */
+  Grid grid_;
+
+  std::size_t num_voxels_;
+
+  std::vector<pcl::PCLPointField> filter_fields_;
+
+  int filter_field_idx_;
+
+  // ============== Below variables are duplicates from GridFilterBase ==============
+  // TODO: maybe declaring friend in GridFilterBase is better?
+  double filter_limit_min_, filter_limit_max_;
+
+  std::string filter_field_name_;
+
+private:
+  friend class ::VoxelGridProtectedMethods_GridFilters_Test;
```

Comment:
namespace pcl::test(s)?

Reply:
`pcl::test` currently is for [test marocs](https://pointclouds.org/documentation/namespacepcl_1_1test.html), and we don't have anything similar unit test in pcl

maybe another name `pcl::unittest`?

Reply:
I removed the friend as we don't need it anymore